chain/store/splitstore/splitstore.go

```diff
@@ -0,0 +1,414 @@
+package splitstore
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/bmatsuo/lmdb-go/lmdb"
+
+	blocks "github.com/ipfs/go-block-format"
+	cid "github.com/ipfs/go-cid"
+	dstore "github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+	bstore "github.com/filecoin-project/lotus/lib/blockstore"
+)
+
+const CompactionThreshold = 5 * build.Finality
+
+var baseEpochKey = dstore.NewKey("baseEpoch")
+
+var log = logging.Logger("splitstore")
+
+type SplitStore struct {
+	baseEpoch abi.ChainEpoch
+	curTs     *types.TipSet
+
+	cs *store.ChainStore
+	ds dstore.Datastore
+
+	hot  bstore.Blockstore
+	cold bstore.Blockstore
+
+	snoop TrackingStore
+
+	stateMx    sync.Mutex
+	compacting bool
+
+	env *lmdb.Env
+}
+
+var _ bstore.Blockstore = (*SplitStore)(nil)
+
+// Blockstore interface
+func (s *SplitStore) DeleteBlock(cid cid.Cid) error {
+	// afaict we don't seem to be using this method, so it's not implemented
+	return errors.New("DeleteBlock not implemented on SplitStore; don't do this Luke!") //nolint
+}
+
+func (s *SplitStore) Has(cid cid.Cid) (bool, error) {
+	has, err := s.hot.Has(cid)
+
+	if err != nil {
+		return false, err
+	}
+
+	if has {
+		return true, nil
+	}
+
+	return s.cold.Has(cid)
+}
+
+func (s *SplitStore) Get(cid cid.Cid) (blocks.Block, error) {
+	blk, err := s.hot.Get(cid)
+
+	switch err {
+	case nil:
+		return blk, nil
+
+	case bstore.ErrNotFound:
+		return s.cold.Get(cid)
+
+	default:
+		return nil, err
+	}
+}
+
+func (s *SplitStore) GetSize(cid cid.Cid) (int, error) {
+	size, err := s.hot.GetSize(cid)
+
+	switch err {
+	case nil:
+		return size, nil
+
+	case bstore.ErrNotFound:
+		return s.cold.GetSize(cid)
+
+	default:
+		return 0, err
+	}
+}
+
+func (s *SplitStore) Put(blk blocks.Block) error {
+	epoch := s.curTs.Height()
+	err := s.snoop.Put(blk.Cid(), epoch)
+	if err != nil {
+		return err
```

Comment:
Hum, I'd prefer to have some kind of graceful degradation here if the snooper fails. Basically we should warn loudly that we've degraded, and disable GC/migration entirely when this happens, instead of failing the `Put` and interrupting service.

Reply:
If the tracker fails we can't effectively do compaction and the hotstore will eventually fill leading to catastrophic failure.
I don't see how graceful degradation helps here.

Reply:
We cannot stop the world if tracking a CID fails. When that happens, we should disable GC entirely, and keep everything in the hot store. It's a safe fallback instead of a hard failure.

Reply:
Writing to the hotstore then is a mistake -- it will lead to catastrophic failure further down the line or result in a leak because of an untracked CID.
If we are to write somewhere, that should be the coldstore.

Reply:
I set it to log the error and fallback to the coldstore.